File: artifacts/api-server/src/routes/tickets.ts
import { Router, type IRouter } from "express";
import { db } from "@workspace/db";
import {
  ticketsTable,
  ticketNotesTable,
  auditLogsTable,
  agentsTable,
  aiResolutionsTable,
} from "@workspace/db";
import { eq, and, desc, or, ilike, sql, gte, lte } from "drizzle-orm";

const router: IRouter = Router();

function buildTicketResponse(ticket: any, agentName?: string | null) {
  return {
    ticket_id: ticket.ticket_id,
    freshservice_ticket_id: ticket.freshservice_ticket_id ?? null,
    title: ticket.title,
    description: ticket.description ?? null,
    use_case: ticket.use_case,
    status: ticket.status,
    priority: ticket.priority,
    sla_deadline: ticket.sla_deadline ? ticket.sla_deadline.toISOString() : null,
    sla_status: ticket.sla_status,
    sla_breach_predicted: ticket.sla_breach_predicted,
    source: ticket.source,
    assigned_agent_id: ticket.assigned_agent_id ?? null,
    assigned_agent_name: agentName ?? null,
    resolution_type: ticket.resolution_type ?? null,
    confidence_score: ticket.confidence_score ? parseFloat(ticket.confidence_score) : null,
    created_at: ticket.created_at.toISOString(),
    updated_at: ticket.updated_at ? ticket.updated_at.toISOString() : null,
    closed_at: ticket.closed_at ? ticket.closed_at.toISOString() : null,
  };
}

router.get("/v1/tickets", async (req, res): Promise<void> => {
  const { status, use_case, priority, sla_status, search, date_from, date_to, page = "1", limit = "20" } = req.query as Record<string, string>;

  const pageNum = parseInt(page, 10) || 1;
  const limitNum = Math.min(parseInt(limit, 10) || 20, 100);
  const offset = (pageNum - 1) * limitNum;

  const conditions: ReturnType<typeof eq>[] = [];
  if (status) conditions.push(eq(ticketsTable.status, status as any));
  if (use_case) conditions.push(eq(ticketsTable.use_case, use_case as any));
  if (priority) conditions.push(eq(ticketsTable.priority, priority as any));
  if (sla_status) conditions.push(eq(ticketsTable.sla_status, sla_status as any));
  if (date_from) conditions.push(gte(ticketsTable.created_at, new Date(date_from)));
  if (date_to) conditions.push(lte(ticketsTable.created_at, new Date(date_to)));

  const whereClause = conditions.length > 0 ? and(...conditions) : undefined;

  const [tickets, countResult] = await Promise.all([
    db.select().from(ticketsTable).where(whereClause).orderBy(desc(ticketsTable.created_at)).limit(limitNum).offset(offset),
    db.select({ count: sql<number>`count(*)` }).from(ticketsTable).where(whereClause),
  ]);

  const agentIds = tickets.map(t => t.assigned_agent_id).filter(Boolean) as string[];
  const agentMap: Record<string, string> = {};
  if (agentIds.length > 0) {
    const agents = await db.select({ agent_id: agentsTable.agent_id, full_name: agentsTable.full_name }).from(agentsTable);
    agents.forEach(a => { agentMap[a.agent_id] = a.full_name; });
  }

  res.json({
    tickets: tickets.map(t => buildTicketResponse(t, t.assigned_agent_id ? agentMap[t.assigned_agent_id] : null)),
    total: Number(countResult[0]?.count ?? 0),
    page: pageNum,
    limit: limitNum,
  });
});

router.post("/v1/tickets", async (req, res): Promise<void> => {
  const { title, description, use_case, priority = "medium", source } = req.body;

  if (!title || !use_case || !source) {
    res.status(400).json({ error: "title, use_case, and source are required" });
    return;
  }

  const [ticket] = await db.insert(ticketsTable).values({
    title,
    description,
    use_case,
    priority,
    source,
    status: "open",
    sla_status: "safe",
    sla_breach_predicted: false,
  }).returning();

  await db.insert(auditLogsTable).values({
    ticket_id: ticket.ticket_id,
    event_type: "ticket_created",
    actor: "system",
    actor_type: "system",
    details: { title, use_case, source },
  });

  res.status(201).json(buildTicketResponse(ticket));
});

router.get("/v1/tickets/:id", async (req, res): Promise<void> => {
  const id = Array.isArray(req.params.id) ? req.params.id[0] : req.params.id;

  const [ticket] = await db.select().from(ticketsTable).where(eq(ticketsTable.ticket_id, id));
  if (!ticket) {
    res.status(404).json({ error: "Ticket not found" });
    return;
  }

  let agentName: string | null = null;
  if (ticket.assigned_agent_id) {
    const [agent] = await db.select().from(agentsTable).where(eq(agentsTable.agent_id, ticket.assigned_agent_id));
    agentName = agent?.full_name ?? null;
  }

  const [aiRes] = await db.select().from(aiResolutionsTable).where(eq(aiResolutionsTable.ticket_id, id)).orderBy(desc(aiResolutionsTable.created_at)).limit(1);
  const notes = await db.select().from(ticketNotesTable).where(eq(ticketNotesTable.ticket_id, id)).orderBy(desc(ticketNotesTable.created_at));

  const base = buildTicketResponse(ticket, agentName);
  res.json({
    ...base,
    ai_resolution: aiRes ? {
      resolution_id: aiRes.resolution_id,
      ticket_id: aiRes.ticket_id,
      intent_detected: aiRes.intent_detected ?? null,
      root_cause: aiRes.root_cause ?? null,
      confidence_score: aiRes.confidence_score,
      intent_clarity_score: aiRes.intent_clarity_score ?? null,
      sop_match_score: aiRes.sop_match_score ?? null,
      historical_success_score: aiRes.historical_success_score ?? null,
      input_completeness_score: aiRes.input_completeness_score ?? null,
      decision: aiRes.decision,
      resolution_steps: (aiRes.resolution_steps as string[]) ?? null,
      execution_status: aiRes.execution_status ?? null,
      execution_output: aiRes.execution_output ?? null,
      time_taken_seconds: aiRes.time_taken_seconds ?? null,
      created_at: aiRes.created_at.toISOString(),
    } : null,
    notes: notes.map(n => ({
      note_id: n.note_id,
      ticket_id: n.ticket_id,
      note_type: n.note_type,
      content: n.content,
      created_by: n.created_by ?? null,
      created_at: n.created_at.toISOString(),
    })),
  });
});

router.patch("/v1/tickets/:id", async (req, res): Promise<void> => {
  const id = Array.isArray(req.params.id) ? req.params.id[0] : req.params.id;
  const { status, priority, assigned_agent_id } = req.body;

  const updateData: any = { updated_at: new Date() };
  if (status) updateData.status = status;
  if (priority) updateData.priority = priority;
  if (assigned_agent_id !== undefined) updateData.assigned_agent_id = assigned_agent_id;
  if (status === "closed") updateData.closed_at = new Date();

  const [ticket] = await db.update(ticketsTable).set(updateData).where(eq(ticketsTable.ticket_id, id)).returning();
  if (!ticket) {
    res.status(404).json({ error: "Ticket not found" });
    return;
  }

  res.json(buildTicketResponse(ticket));
});

router.delete("/v1/tickets/:id", async (req, res): Promise<void> => {
  const id = Array.isArray(req.params.id) ? req.params.id[0] : req.params.id;

  const [ticket] = await db.delete(ticketsTable).where(eq(ticketsTable.ticket_id, id)).returning();
  if (!ticket) {
    res.status(404).json({ error: "Ticket not found" });
    return;
  }

  res.json({ success: true, message: "Ticket deleted" });
});

router.post("/v1/tickets/:id/resolve", async (req, res): Promise<void> => {
  const id = Array.isArray(req.params.id) ? req.params.id[0] : req.params.id;

  const [ticket] = await db.select().from(ticketsTable).where(eq(ticketsTable.ticket_id, id));
  if (!ticket) {
    res.status(404).json({ error: "Ticket not found" });
    return;
  }

  const intentClarity = 0.75 + Math.random() * 0.2;
  const sopMatch = 0.70 + Math.random() * 0.25;
  const historicalSuccess = 0.80;
  const inputCompleteness = 0.90;
  const confidence = (intentClarity * 0.30 + sopMatch * 0.35 + historicalSuccess * 0.25 + inputCompleteness * 0.10) * 100;

  let decision: "auto_resolve" | "review_after" | "escalate" = "escalate";
  if (confidence >= 85) decision = "auto_resolve";
  else if (confidence >= 60) decision = "review_after";

  const steps = [
    `Analyzed ticket: "${ticket.title}"`,
    `Intent detected: ${ticket.use_case.replace(/_/g, " ")}`,
    `SOP match score: ${(sopMatch * 100).toFixed(1)}%`,
    decision === "auto_resolve" ? "Auto-resolution initiated successfully" : "Escalation recommended based on confidence score",
  ];

  const [resolution] = await db.insert(aiResolutionsTable).values({
    ticket_id: id,
    intent_detected: ticket.use_case.replace(/_/g, " "),
    root_cause: `AI analysis of: ${ticket.title}`,
    confidence_score: confidence / 100,
    intent_clarity_score: intentClarity,
    sop_match_score: sopMatch,
    historical_success_score: historicalSuccess,
    input_completeness_score: inputCompleteness,
    decision,
    resolution_steps: steps,
    execution_status: decision === "auto_resolve" ? "success" : "partial",
    execution_output: decision === "auto_resolve" ? "Resolution completed successfully." : "Requires human review.",
    time_taken_seconds: Math.floor(Math.random() * 30) + 5,
  }).returning();

  const newStatus = decision === "auto_resolve" ? "auto_resolved" : decision === "escalate" ? "escalated" : "in_progress";
  await db.update(ticketsTable).set({
    status: newStatus,
    confidence_score: String(confidence / 100),
    resolution_type: decision === "auto_resolve" ? "auto" : "manual",
    updated_at: new Date(),
    ...(newStatus === "auto_resolved" ? { closed_at: new Date() } : {}),
  }).where(eq(ticketsTable.ticket_id, id));

  await db.insert(auditLogsTable).values({
    ticket_id: id,
    event_type: decision === "auto_resolve" ? "auto_resolved" : "escalated",
    actor: "STACK AI",
    actor_type: "ai",
    details: { confidence: confidence.toFixed(1), decision },
  });

  res.json({
    resolution_id: resolution.resolution_id,
    ticket_id: resolution.ticket_id,
    intent_detected: resolution.intent_detected,
    root_cause: resolution.root_cause,
    confidence_score: resolution.confidence_score,
    intent_clarity_score: resolution.intent_clarity_score,
    sop_match_score: resolution.sop_match_score,
    historical_success_score: resolution.historical_success_score,
    input_completeness_score: resolution.input_completeness_score,
    decision: resolution.decision,
    resolution_steps: resolution.resolution_steps,
    execution_status: resolution.execution_status,
    execution_output: resolution.execution_output,
    time_taken_seconds: resolution.time_taken_seconds,
    created_at: resolution.created_at.toISOString(),
  });
});

router.post("/v1/tickets/:id/escalate", async (req, res): Promise<void> => {
  const id = Array.isArray(req.params.id) ? req.params.id[0] : req.params.id;
  const { reason } = req.body;

  const [ticket] = await db.update(ticketsTable).set({ status: "escalated", updated_at: new Date() }).where(eq(ticketsTable.ticket_id, id)).returning();
  if (!ticket) {
    res.status(404).json({ error: "Ticket not found" });
    return;
  }

  await db.insert(auditLogsTable).values({
    ticket_id: id,
    event_type: "escalated",
    actor: "agent",
    actor_type: "agent",
    details: { reason },
  });

  res.json({ success: true, message: "Ticket escalated" });
});

router.get("/v1/tickets/:id/notes", async (req, res): Promise<void> => {
  const id = Array.isArray(req.params.id) ? req.params.id[0] : req.params.id;

  const notes = await db.select().from(ticketNotesTable).where(eq(ticketNotesTable.ticket_id, id)).orderBy(desc(ticketNotesTable.created_at));
  res.json(notes.map(n => ({
    note_id: n.note_id,
    ticket_id: n.ticket_id,
    note_type: n.note_type,
    content: n.content,
    created_by: n.created_by ?? null,
    created_at: n.created_at.toISOString(),
  })));
});

router.post("/v1/tickets/:id/notes", async (req, res): Promise<void> => {
  const id = Array.isArray(req.params.id) ? req.params.id[0] : req.params.id;
  const { content, note_type = "human_note" } = req.body;

  if (!content) {
    res.status(400).json({ error: "content is required" });
    return;
  }

  const [note] = await db.insert(ticketNotesTable).values({
    ticket_id: id,
    content,
    note_type,
    created_by: "agent",
  }).returning();

  res.status(201).json({
    note_id: note.note_id,
    ticket_id: note.ticket_id,
    note_type: note.note_type,
    content: note.content,
    created_by: note.created_by ?? null,
    created_at: note.created_at.toISOString(),
  });
});

router.get("/v1/tickets/:id/timeline", async (req, res): Promise<void> => {
  const id = Array.isArray(req.params.id) ? req.params.id[0] : req.params.id;

  const logs = await db.select().from(auditLogsTable).where(eq(auditLogsTable.ticket_id, id)).orderBy(desc(auditLogsTable.created_at));
  res.json(logs.map(l => ({
    log_id: l.log_id,
    ticket_id: l.ticket_id ?? null,
    event_type: l.event_type,
    actor: l.actor,
    actor_type: l.actor_type,
    details: l.details ?? {},
    ip_address: l.ip_address ?? null,
    created_at: l.created_at.toISOString(),
  })));
});

export default router;

File: artifacts/api-server/app/routers/tickets.py
import uuid
from datetime import datetime, timedelta
from typing import Optional
from fastapi import APIRouter, Depends, HTTPException, Query, BackgroundTasks
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, func, or_
from sqlalchemy.orm import selectinload

from app.database import get_db
from app.models.tickets import Ticket, TicketNote
from app.models.agents import Agent
from app.models.logs import AuditLog
from app.schemas.tickets import (
    TicketOut, TicketCreate, TicketUpdate, TicketNoteCreate,
    TicketNoteOut, TicketResolveRequest, TicketEscalateRequest,
)
from app.middleware.auth import get_current_agent

router = APIRouter()


async def _log_event(db, ticket_id, event_type, actor, actor_type="agent", details=None):
    log = AuditLog(
        ticket_id=ticket_id,
        event_type=event_type,
        actor=actor,
        actor_type=actor_type,
        details=details or {},
    )
    db.add(log)


@router.get("")
async def list_tickets(
    status: Optional[str] = None,
    priority: Optional[str] = None,
    use_case: Optional[str] = None,
    sla_status: Optional[str] = None,
    search: Optional[str] = None,
    date_from: Optional[str] = None,
    date_to: Optional[str] = None,
    page: int = Query(1, ge=1),
    limit: int = Query(20, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
    agent: Agent = Depends(get_current_agent),
):
    q = select(Ticket).options(
        selectinload(Ticket.assigned_agent),
        selectinload(Ticket.ai_resolutions),
    )
    if status:
        q = q.where(Ticket.status == status)
    if priority:
        q = q.where(Ticket.priority == priority)
    if use_case:
        q = q.where(Ticket.use_case == use_case)
    if sla_status:
        q = q.where(Ticket.sla_status == sla_status)
    if search:
        q = q.where(or_(
            Ticket.title.ilike(f"%{search}%"),
            Ticket.user_email.ilike(f"%{search}%"),
        ))
    if date_from:
        try:
            q = q.where(Ticket.created_at >= datetime.fromisoformat(date_from))
        except ValueError:
            pass
    if date_to:
        try:
            q = q.where(Ticket.created_at <= datetime.fromisoformat(date_to + "T23:59:59"))
        except ValueError:
            pass

    count_q = select(func.count()).select_from(q.subquery())
    total = (await db.execute(count_q)).scalar_one()

    q = q.order_by(Ticket.created_at.desc()).offset((page - 1) * limit).limit(limit)
    result = await db.execute(q)
    tickets = result.scalars().all()

    return {
        "tickets": [_ticket_to_dict(t) for t in tickets],
        "total": total,
        "page": page,
        "limit": limit,
    }


@router.post("", status_code=201)
async def create_ticket(
    body: TicketCreate,
    background_tasks: BackgroundTasks,
    db: AsyncSession = Depends(get_db),
    agent: Agent = Depends(get_current_agent),
):
    from app.services.resolution.engine import compute_sla_deadline
    deadline = compute_sla_deadline(body.use_case, body.priority)

    ticket = Ticket(
        title=body.title,
        description=body.description,
        use_case=body.use_case,
        priority=body.priority or "medium",
        source=body.source or "web_dashboard",
        user_email=body.user_email,
        assigned_agent_id=body.assigned_agent_id,
        sla_deadline=deadline,
        sla_status="safe",
    )
    db.add(ticket)
    await db.flush()
    await _log_event(db, ticket.ticket_id, "ticket_created", agent.email, "agent",
                     {"title": ticket.title, "use_case": ticket.use_case})

    background_tasks.add_task(_trigger_ai_triage, str(ticket.ticket_id))

    await db.refresh(ticket, ["assigned_agent", "ai_resolutions"])
    return _ticket_to_dict(ticket)


async def _trigger_ai_triage(ticket_id: str):
    from app.services.ai.langchain_agent import triage_ticket
    try:
        await triage_ticket(ticket_id)
    except Exception:
        pass


@router.get("/{ticket_id}")
async def get_ticket(
    ticket_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    agent: Agent = Depends(get_current_agent),
):
    result = await db.execute(
        select(Ticket).options(
            selectinload(Ticket.assigned_agent),
            selectinload(Ticket.notes),
            selectinload(Ticket.ai_resolutions),
            selectinload(Ticket.attachments),
        ).where(Ticket.ticket_id == ticket_id)
    )
    ticket = result.scalar_one_or_none()
    if not ticket:
        raise HTTPException(status_code=404, detail="Ticket not found")
    return _ticket_detail_to_dict(ticket)


@router.patch("/{ticket_id}")
async def update_ticket(
    ticket_id: uuid.UUID,
    body: TicketUpdate,
    db: AsyncSession = Depends(get_db),
    agent: Agent = Depends(get_current_agent),
):
    result = await db.execute(
        select(Ticket).options(
            selectinload(Ticket.assigned_agent),
            selectinload(Ticket.ai_resolutions),
            selectinload(Ticket.notes),
        ).where(Ticket.ticket_id == ticket_id)
    )
    ticket = result.scalar_one_or_none()
    if not ticket:
        raise HTTPException(status_code=404, detail="Ticket not found")

    changes = {}
    for field, val in body.model_dump(exclude_none=True).items():
        setattr(ticket, field, val)
        changes[field] = val

    if body.status in ("auto_resolved", "closed"):
        ticket.closed_at = datetime.utcnow()

    await _log_event(db, ticket_id, "ticket_updated", agent.email, "agent", changes)
    return _ticket_to_dict(ticket)


@router.delete("/{ticket_id}")
async def delete_ticket(
    ticket_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    agent: Agent = Depends(get_current_agent),
):
    result = await db.execute(select(Ticket).where(Ticket.ticket_id == ticket_id))
    ticket = result.scalar_one_or_none()
    if not ticket:
        raise HTTPException(status_code=404, detail="Ticket not found")
    await db.delete(ticket)
    return {"success": True, "message": "Ticket deleted"}


@router.post("/{ticket_id}/resolve")
async def resolve_ticket(
    ticket_id: uuid.UUID,
    body: TicketResolveRequest,
    background_tasks: BackgroundTasks,
    db: AsyncSession = Depends(get_db),
    agent: Agent = Depends(get_current_agent),
):
    result = await db.execute(
        select(Ticket).options(
            selectinload(Ticket.assigned_agent),
            selectinload(Ticket.ai_resolutions),
        ).where(Ticket.ticket_id == ticket_id)
    )
    ticket = result.scalar_one_or_none()
    if not ticket:
        raise HTTPException(status_code=404, detail="Ticket not found")

    ticket.status = "auto_resolved"
    ticket.resolution_type = "manual" if body.force else "auto"
    ticket.closed_at = datetime.utcnow()

    if body.resolution_notes:
        note = TicketNote(
            ticket_id=ticket_id,
            note_type="resolution_summary",
            content=body.resolution_notes,
            created_by=agent.email,
        )
        db.add(note)

    await _log_event(db, ticket_id, "ticket_resolved", agent.email, "agent",
                     {"force": body.force, "resolution_type": ticket.resolution_type})
    background_tasks.add_task(_sync_to_freshservice, str(ticket_id), "resolved")
    return _ticket_to_dict(ticket)


@router.post("/{ticket_id}/escalate")
async def escalate_ticket(
    ticket_id: uuid.UUID,
    body: TicketEscalateRequest,
    db: AsyncSession = Depends(get_db),
    agent: Agent = Depends(get_current_agent),
):
    result = await db.execute(
        select(Ticket).options(
            selectinload(Ticket.assigned_agent),
            selectinload(Ticket.ai_resolutions),
        ).where(Ticket.ticket_id == ticket_id)
    )
    ticket = result.scalar_one_or_none()
    if not ticket:
        raise HTTPException(status_code=404, detail="Ticket not found")

    ticket.status = "escalated"
    ticket.priority = "high" if ticket.priority == "medium" else ticket.priority

    if body.reason:
        note = TicketNote(
            ticket_id=ticket_id,
            note_type="ai_context",
            content=f"Escalated: {body.reason}",
            created_by="system",
        )
        db.add(note)

    await _log_event(db, ticket_id, "ticket_escalated", agent.email, "agent",
                     {"reason": body.reason, "target_group": body.target_group})
    return _ticket_to_dict(ticket)


@router.get("/{ticket_id}/notes", response_model=list[TicketNoteOut])
async def get_notes(
    ticket_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    agent: Agent = Depends(get_current_agent),
):
    result = await db.execute(
        select(TicketNote).where(TicketNote.ticket_id == ticket_id).order_by(TicketNote.created_at)
    )
    return [TicketNoteOut.model_validate(n) for n in result.scalars().all()]


@router.post("/{ticket_id}/notes", response_model=TicketNoteOut, status_code=201)
async def add_note(
    ticket_id: uuid.UUID,
    body: TicketNoteCreate,
    db: AsyncSession = Depends(get_db),
    agent: Agent = Depends(get_current_agent),
):
    note = TicketNote(
        ticket_id=ticket_id,
        note_type=body.note_type or "human_note",
        content=body.content,
        created_by=body.created_by or agent.email,
    )
    db.add(note)
    await db.flush()
    return TicketNoteOut.model_validate(note)


@router.get("/{ticket_id}/timeline")
async def get_timeline(
    ticket_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    agent: Agent = Depends(get_current_agent),
):
    result = await db.execute(
        select(AuditLog).where(AuditLog.ticket_id == ticket_id).order_by(AuditLog.created_at)
    )
    logs = result.scalars().all()
    return [
        {
            "log_id": str(l.log_id),
            "event_type": l.event_type,
            "actor": l.actor,
            "actor_type": l.actor_type,
            "details": l.details,
            "created_at": l.created_at.isoformat() if l.created_at else None,
        }
        for l in logs
    ]


def _ticket_to_dict(t: Ticket) -> dict:
    conf = t.confidence_score_float if hasattr(t, 'confidence_score_float') else None
    return {
        "ticket_id": str(t.ticket_id),
        "freshservice_ticket_id": t.freshservice_ticket_id,
        "title": t.title,
        "description": t.description,
        "use_case": t.use_case,
        "status": t.status,
        "priority": t.priority,
        "sla_deadline": t.sla_deadline.isoformat() if t.sla_deadline else None,
        "sla_status": t.sla_status,
        "sla_breach_predicted": t.sla_breach_predicted or False,
        "source": t.source,
        "user_email": t.user_email,
        "assigned_agent_id": str(t.assigned_agent_id) if t.assigned_agent_id else None,
        "assigned_agent_name": t.assigned_agent.full_name if t.assigned_agent else None,
        "resolution_type": t.resolution_type,
        "confidence_score": conf,
        "created_at": t.created_at.isoformat() if t.created_at else None,
        "updated_at": t.updated_at.isoformat() if t.updated_at else None,
        "closed_at": t.closed_at.isoformat() if t.closed_at else None,
    }


def _ticket_detail_to_dict(t: Ticket) -> dict:
    base = _ticket_to_dict(t)
    ai_res = None
    if t.ai_resolutions:
        r = t.ai_resolutions[-1]
        ai_res = {
            "resolution_id": str(r.resolution_id),
            "ticket_id": str(r.ticket_id),
            "intent_detected": r.intent_detected,
            "root_cause": r.root_cause,
            "confidence_score": r.confidence_score or 0,
            "intent_clarity_score": r.intent_clarity_score,
            "sop_match_score": r.sop_match_score,
            "historical_success_score": r.historical_success_score,
            "input_completeness_score": r.input_completeness_score,
            "decision": r.decision,
            "resolution_steps": r.resolution_steps or [],
            "execution_status": r.execution_status,
            "execution_output": r.execution_output,
            "time_taken_seconds": r.time_taken_seconds,
            "created_at": r.created_at.isoformat() if r.created_at else None,
        }
    base["ai_resolution"] = ai_res
    base["notes"] = [
        {
            "note_id": str(n.note_id),
            "ticket_id": str(n.ticket_id),
            "note_type": n.note_type,
            "content": n.content,
            "created_by": n.created_by,
            "created_at": n.created_at.isoformat() if n.created_at else None,
        }
        for n in (t.notes or [])
    ]
    return base


async def _sync_to_freshservice(ticket_id: str, status: str):
    from app.services.integrations.freshservice import FreshserviceClient
    try:
        client = FreshserviceClient()
        await client.update_ticket_status(ticket_id, status)
    except Exception:
        pass
